File: go/internal/tools/runtime_env.go
package tools

import (
	"context"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

type pathEntry struct {
	Path   string
	Source string
}

func (e *Engine) effectivePath() ([]pathEntry, []string) {
	home, _ := os.UserHomeDir()
	var candidates []pathEntry
	for _, value := range e.settings.MCPExtraPath {
		candidates = append(candidates, pathEntry{value, "explicit_extra"})
	}
	for _, value := range filepath.SplitList(os.Getenv("PATH")) {
		candidates = append(candidates, pathEntry{value, "inherited_path"})
	}
	if value := os.Getenv("VIRTUAL_ENV"); value != "" {
		candidates = append(candidates, pathEntry{filepath.Join(value, "bin"), "virtualenv"})
	}
	for _, item := range []struct{ env, suffix string }{{"GOROOT", "bin"}, {"GOPATH", "bin"}, {"CARGO_HOME", "bin"}, {"PYENV_ROOT", "shims"}} {
		if value := os.Getenv(item.env); value != "" {
			candidates = append(candidates, pathEntry{filepath.Join(value, item.suffix), "standard_fallback"})
		}
	}
	if value := os.Getenv("NVM_BIN"); value != "" {
		candidates = append(candidates, pathEntry{value, "standard_fallback"})
	}
	for _, value := range []string{"/usr/local/go/bin", "/usr/local/bin", "/usr/bin", filepath.Join(home, "go/bin"), filepath.Join(home, ".local/bin"), filepath.Join(home, ".cargo/bin")} {
		candidates = append(candidates, pathEntry{value, "standard_fallback"})
	}
	seen := map[string]bool{}
	entries := make([]pathEntry, 0, len(candidates))
	for _, entry := range candidates {
		if entry.Path == "" {
			continue
		}
		absolute, err := filepath.Abs(entry.Path)
		if err != nil || seen[absolute] {
			continue
		}
		info, err := os.Stat(absolute)
		if err != nil || !info.IsDir() {
			continue
		}
		seen[absolute] = true
		entries = append(entries, pathEntry{absolute, entry.Source})
	}
	if len(entries) == 0 {
		return entries, []string{"effective PATH contains no existing directories"}
	}
	return entries, nil
}

func (e *Engine) commandEnvironment(overrides map[string]string) []string {
	entries, _ := e.effectivePath()
	paths := make([]string, 0, len(entries))
	for _, entry := range entries {
		paths = append(paths, entry.Path)
	}
	environment := make([]string, 0, len(os.Environ())+len(overrides)+1)
	for _, value := range os.Environ() {
		if !strings.HasPrefix(value, "PATH=") {
			environment = append(environment, value)
		}
	}
	environment = append(environment, "PATH="+strings.Join(paths, string(os.PathListSeparator)))
	for key, value := range overrides {
		environment = append(environment, key+"="+value)
	}
	return environment
}

// resolveExecutable uses the same explicit-first PATH contract exposed by
// doctor and inherited by child processes. It intentionally never searches a
// repository-local virtual environment: the standalone server stays portable.
func (e *Engine) resolveExecutable(name string) (string, string) {
	entries, _ := e.effectivePath()
	candidates := []string{name}
	if runtime.GOOS == "windows" && filepath.Ext(name) == "" {
		candidates = append(candidates, name+".exe", name+".cmd", name+".bat")
	}
	for _, entry := range entries {
		for _, candidateName := range candidates {
			candidate := filepath.Join(entry.Path, candidateName)
			if info, err := os.Stat(candidate); err == nil && !info.IsDir() && info.Mode()&0o111 != 0 {
				return candidate, entry.Source
			}
		}
	}
	return "", "not_found"
}

func (e *Engine) toolStatus(name string) map[string]any {
	path, source := e.resolveExecutable(name)
	result := map[string]any{"name": name, "available": path != "", "path": nil, "source": "not_found", "version": nil, "version_error": nil}
	if path == "" {
		return result
	}
	result["path"], result["source"] = path, source
	args := []string{"--version"}
	if name == "go" {
		args = []string{"version"}
	}
	process := runProcess(context.Background(), e.settings.ProjectRoot, 3*time.Second, e.commandEnvironment(nil), path, args...)
	line := strings.Split(strings.TrimSpace(process.Stdout+"\n"+process.Stderr), "\n")[0]
	if process.ExitCode != 0 {
		result["version_error"] = firstNonEmpty(line, "version command failed")
	} else {
		result["version"] = line
	}
	return result
}

File: python/src/chatrepo_mcp/runtime_env.py
from __future__ import annotations

import os
import re
import subprocess
import sys
from pathlib import Path
from typing import Any

from .config import Settings


_VERSION_ARGS = {
    "go": ("version",),
    "node": ("--version",),
    "npm": ("--version",),
    "python3": ("--version",),
    "git": ("--version",),
    "rg": ("--version",),
    "gh": ("--version",),
    "ctags": ("--version",),
    "ruff": ("--version",),
    "mypy": ("--version",),
    "pyright": ("--version",),
}


def _existing_dirs(values: list[tuple[str, str]]) -> tuple[list[str], dict[str, str]]:
    paths: list[str] = []
    sources: dict[str, str] = {}
    seen: set[str] = set()
    for raw, source in values:
        if not raw:
            continue
        path = str(Path(raw).expanduser().resolve())
        if path in seen or not Path(path).is_dir():
            continue
        seen.add(path)
        paths.append(path)
        sources[path] = source
    return paths, sources


def effective_path(settings: Settings) -> tuple[list[str], dict[str, str], list[str]]:
    home = Path.home()
    candidates: list[tuple[str, str]] = [
        *((value, "explicit_extra") for value in settings.mcp_extra_path),
        *((value, "inherited_path") for value in os.environ.get("PATH", "").split(os.pathsep)),
    ]
    virtual_env = os.environ.get("VIRTUAL_ENV")
    if virtual_env:
        candidates.append((str(Path(virtual_env) / "bin"), "virtualenv"))
    executable_bin = str(Path(sys.executable).resolve().parent)
    candidates.append((executable_bin, "virtualenv"))
    for variable, suffix in (
        ("GOROOT", "bin"),
        ("GOPATH", "bin"),
        ("CARGO_HOME", "bin"),
        ("PYENV_ROOT", "shims"),
    ):
        if value := os.environ.get(variable):
            candidates.append((str(Path(value) / suffix), "standard_fallback"))
    if value := os.environ.get("NVM_BIN"):
        candidates.append((value, "standard_fallback"))
    candidates.extend(
        (value, "standard_fallback")
        for value in (
            "/usr/local/go/bin",
            "/usr/local/bin",
            "/usr/bin",
            str(home / "go/bin"),
            str(home / ".local/bin"),
            str(home / ".cargo/bin"),
        )
    )
    paths, sources = _existing_dirs(candidates)
    warnings = [] if paths else ["effective PATH contains no existing directories"]
    return paths, sources, warnings


def command_environment(settings: Settings, overrides: dict[str, str] | None = None) -> dict[str, str]:
    env = os.environ.copy()
    paths, _, _ = effective_path(settings)
    env["PATH"] = os.pathsep.join(paths)
    for key, value in (overrides or {}).items():
        if not re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", key):
            raise ValueError(f"invalid env key: {key}")
        env[key] = value
    return env


def resolve_binary(name: str, settings: Settings) -> tuple[str | None, str]:
    paths, sources, _ = effective_path(settings)
    for directory in paths:
        candidate = Path(directory) / name
        if candidate.is_file() and os.access(candidate, os.X_OK):
            return str(candidate), sources[directory]
    return None, "not_found"


def tool_status(name: str, settings: Settings) -> dict[str, Any]:
    path, source = resolve_binary(name, settings)
    result: dict[str, Any] = {
        "name": name,
        "available": path is not None,
        "path": path,
        "source": source,
        "version": None,
        "version_error": None,
    }
    if path is None:
        return result
    try:
        proc = subprocess.run(
            [path, *_VERSION_ARGS.get(name, ("--version",))],
            env=command_environment(settings),
            text=True,
            capture_output=True,
            timeout=3,
            check=False,
        )
        output = (proc.stdout or proc.stderr).strip().splitlines()
        if proc.returncode == 0 and output:
            result["version"] = output[0][:300]
        else:
            result["version_error"] = (proc.stderr or proc.stdout or f"exit {proc.returncode}").strip()[:300]
    except (OSError, subprocess.SubprocessError) as exc:
        result["version_error"] = str(exc)
    return result
